src/rewriter/README.md

```diff
@@ -0,0 +1,231 @@
+  ((xs ?BitVec :list)
+   (s ?BitVec)
+   (ys ?BitVec :list)
+   (zs ?BitVec :list))
+  (concat xs (concat s ys) zs)
+  (concat xs s ys zs))
+```
+
+The user can optionally supply a rewrite context to support the continuation.
+The context indicates how to use the recursion step, which rewrites *match* to
+*target*, to construct the final result. Below is an example
+
+``` lisp
+(define-rule* str-len-concat-rec ((s1 String) (s2 String) (s3 String :list))
+	(str.len (str.++ s1 s2 s3))
+	(str.len (str.++ s2 s3))
+	(+ (str.len s1) _))
+```
+
+This rule specifies that the rewrite of `(str.len (str.++ s1 s2 ...))` is the
+sum of `(str.len s1)` and the result of applying the rule to the term `(str.len
+(str.++ s2 ...))`
+
+
+## Adding New Operators
+
+This section pertains only to parsing RARE rules in cvc5.
+
+The RARE parser is located at `src/rewriter`. To add a new operator,
+write a new enumeration entry in `src/rewriter/node.py`. The syntax and
+processing are controlled by `rewriter.py` and `parser.py`.
+
+The symbols usable in `(<id> <expr>+)` can be found in `node.py`. The sorts
+available can be found in `mkrewrites.py`.
+
+It is possible in the rewriter to enable parsing of rules of the form
+`define-cond-rule*`. However the condition may interact unexpectedly with the
+fixed-point rule.
+
+```
+<rule> ::= (define-cond-rule* <symbol> (<par>*) [<defs>] <expr> <expr> <expr> [<expr>])
+```
+
+## Changes from Previous Works
+
+In comparison to
+[fmcad22](https://ieeexplore.ieee.org/abstract/document/10026573), we have
+removed the `let` expression.  `let` is replaced by `def` which allow symbols to
+be shared across the condition, match, and target terms.
+
+We also deleted the `const` modifier, which existed in some previous works and
+in past versions of the parser. It is not necessary anymore to flag parameters
+as constants. The rationale is that there used to be 3 reasons to use `:const`:
+`:const`:
+1. There should not be a rule that is only sound when its arguments are
+   constant, so `:const` should not exist in this case.
+2. The evaluation of operator applications of all constant arguments is handled
+   by the rewriter, and therefore rules evaluating constant expressions should
+   not exist in RARE
+3. There are some conditional rules in `theory/bv` which only apply when the
+   arguments satisfy a condition. In this case, we could match the required
+   argument using a `(bv v n)` term with variable bit-width `n` instead of an
+   abstract bit-vector term `(y ?BitVec)`. For example,
+
+``` lisp
+(define-cond-rule bv-udiv-pow2-1p
+  ((x ?BitVec) (v Int) (n Int))
+  (def (power (int.log2 v)))
+  (and (int.ispow2 v) (> v 1))
+  (bvudiv x (bv v n))
+  (concat (bv 0 power) (extract (- n 1) power x)))
+```
+
+More information can be found in other works but beware of changes:
+* [FMCAD 22/Reconstructing Fine-Grained Proofs of Rewrites Using a Domain-Specific Language](https://ieeexplore.ieee.org/document/10026573)
```

Comment:
```suggestion
```

Reply:
Maybe also convert above changes to a bullet list:

* let ...
* const ...
   1. ...
   2. ...
   3. ...

Reply:
fixed